infer/src/bufferoverrun/bufferOverrunSemantics.ml

```diff
@@ -411,7 +411,24 @@ let rec eval_sympath_partial ~mode params p mem =
         Mem.find (Loc.of_allocsite (Allocsite.make_symbol p)) mem
     | EvalPOCond | EvalPOReachability ->
         Val.Itv.top )
-  | BoField.(Prim (Symb.SymbolPath.Deref _) | Field _ | StarField _) ->
+  | BoField.(Prim (Symb.SymbolPath.Deref _)) ->
+      let locs = eval_locpath ~mode params p mem in
+      PowLoc.fold
+        (fun loc acc ->
+          Val.join
+            (let v = Mem.find loc mem in
+             if Val.is_default v then
+               (* There is no or default value for the dereference of a pointer in the abstract
+                  state of the caller. Create symbolic value representing the dereference.
+                  This makes it possible to import, e.g., the fact that "**A_Callee=2" for
+                  a pointer "*A_Caller" having an unknown value in the caller (A_Callee
+                  being a formal parameter of type pointer to pointer, A_Caller being an actual
+                  parameter of type pointer to pointer). *)
+               Mem.on_demand ~loc mem
+             else v )
```

Comment:
* We may want to use `Mem.find_opt`?

* (If not) When the result of `Mem.find` is a default value, it is calling `Mem.on_demand ~loc mem`. But, IIUC, `Mem.find` already calls `Val.on_demand` in it (as below), and returns the `default` when the `on_demand` failed. So, I am confused about how calling `Mem.on_demand` again makes any difference.

https://github.com/facebook/infer/blob/51820faf5a553653559ce52845979153982c59e4/infer/src/bufferoverrun/bufferOverrunDomain.ml#L2171-L2176

Reply:
Thanks for noticing this, I indeed messed up the code when implementing the switch to chose the default value (top / bottom).

The condition should be `if Val.top v then`. The interesting case is when the dereferenced pointer is present in the abstract state and its value is top. Then, `Val.on_demand` is not called in `Mem.find` and we would get `Loc.Unknown` as a location of its dereference. With the change, we call `Val.on_demand` and get a symbolic value representing the dereference. That is, it is supposed to handle an example like this:

```c
void callee (int** p) {
  **p = 2;
}

void test () {
  int* ptr;
  ptr = havoc; // ptr has unknown value
  callee (&ptr); // we want to import *ptr = 2
}
```

However, I decided to replace this code in the new commit "Make the import of dereferences more robust." that I added to this PR. The new code does basically the same thing but in a less obscure way and it is also more robust - in same cases, because of the lack of type information, `Val.on_demand` would return `Val.Unknown` resulting in returning and unknown location representing the imported dereference while we are able to create a symbolic location representing the dereference in the new code.

Reply:
It sounds great!